### PR TITLE
Made event submission synchronous to avoid missed events.

### PR DESCRIPTION
**Pull request by ibuenros**

### CODE REVIEW

**Comment by ibuenros** [comment]
@pcadabam can you review?

File: gobblin-metrics/src/main/java/gobblin/metrics/MetricContext.java [hunk 1]
```diff
@@ -663,9 +665,31 @@ public static Builder builder(String name) {
...
+    long attemptsLeft = 10;
+
+    while(attemptsLeft > 0) {
+      long key = rand.nextLong();
+      if(this.notificationTargets.putIfAbsent(key, target) == null) {
+        return key;
+      }
+      attemptsLeft--;
+    }
+
```
Comment (pcadabam) [inline gobblin-metrics/src/main/java/gobblin/metrics/MetricContext.java]: Use a UUID instead

**Comment by pcadabam** [comment]
LGTM

File: gobblin-metrics/src/main/java/gobblin/metrics/MetricContext.java [hunk 2]
```diff
@@ -663,9 +665,25 @@ public static Builder builder(String name) {
...
    * @param target A {@link com.google.common.base.Function} that will be run every time
    *               there is a new {@link gobblin.metrics.notification.Notification} in this context.
+   * @return a key for this notification target. Can be used to remove the notification target later.
    */
-  public void addNotificationTarget(Function<Notification, Void> target) {
-    this.notificationTargets.add(target);
+  public UUID addNotificationTarget(Function<Notification, Void> target) {
+
+    UUID uuid = UUID.randomUUID();
+    if(this.notificationTargets.putIfAbsent(uuid, target) == null) {
```
Comment (zliu41) [inline gobblin-metrics/src/main/java/gobblin/metrics/MetricContext.java]: @ibuenros `Map.putIfAbsent()` is not available in Java 6 and 7.

**Pull request merged**